File: src/test/java/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/CommandInvokableReturnTypeInspectionTest.java
package fr.adrienbrault.idea.symfony2plugin.tests.codeInspection.command;

import fr.adrienbrault.idea.symfony2plugin.tests.SymfonyLightCodeInsightFixtureTestCase;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see fr.adrienbrault.idea.symfony2plugin.codeInspection.command.CommandInvokableReturnTypeInspection
 */
public class CommandInvokableReturnTypeInspectionTest extends SymfonyLightCodeInsightFixtureTestCase {
    public void setUp() throws Exception {
        super.setUp();
        myFixture.copyFileToProject("classes.php");
    }

    public String getTestDataPath() {
        return "src/test/java/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/fixtures";
    }

    public void testCommandWithoutIntReturnType() {
        assertLocalInspectionContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __inv<caret>oke(): void\n" +
            "    {\n" +
            "    }\n" +
            "}",
            "Symfony: Consider adding int return type to command __invoke()"
        );
    }

    public void testCommandWithStringReturnType() {
        assertLocalInspectionContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __inv<caret>oke(): string\n" +
            "    {\n" +
            "        return 'test';\n" +
            "    }\n" +
            "}",
            "Symfony: Consider adding int return type to command __invoke()"
        );
    }

    public void testCommandWithNoReturnType() {
        assertLocalInspectionContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __inv<caret>oke()\n" +
            "    {\n" +
            "    }\n" +
            "}",
            "Symfony: Consider adding int return type to command __invoke()"
        );
    }

    public void testCommandWithIntReturnType() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Attribute\\AsCommand;\n" +
            "\n" +
            "#[AsCommand(name: 'app:foobar')]\n" +
            "class FoobarCommand\n" +
            "{\n" +
            "    public function __inv<caret>oke(): int\n" +
            "    {\n" +
            "        return 0;\n" +
            "    }\n" +
            "}",
            "Symfony: Consider adding int return type to command __invoke()"
        );
    }

    public void testNonCommandClassIsIgnored() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "class FoobarClass\n" +
            "{\n" +
            "    public function __inv<caret>oke(): void\n" +
            "    {\n" +
            "    }\n" +
            "}",
            "Symfony: Consider adding int return type to command __invoke()"
        );
    }

    public void testTraditionalCommandWithoutAttributeIsIgnored() {
        assertLocalInspectionNotContains("test.php", "<?php\n" +
            "use Symfony\\Component\\Console\\Command\\Command;\n" +
            "\n" +
            "class FoobarCommand extends Command\n" +
            "{\n" +
            "    public function __inv<caret>oke(): void\n" +
            "    {\n" +
            "    }\n" +
            "}",
            "Symfony: Consider adding int return type to command __invoke()"
        );
    }
}

File: src/test/kotlin/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/CommandInvokableReturnTypeInspectionTest.kt
package fr.adrienbrault.idea.symfony2plugin.tests.codeInspection.command

import fr.adrienbrault.idea.symfony2plugin.codeInspection.command.CommandInvokableReturnTypeInspection
import fr.adrienbrault.idea.symfony2plugin.tests.SymfonyLightCodeInsightFixtureTestCase

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see CommandInvokableReturnTypeInspection
 */
class CommandInvokableReturnTypeInspectionTest : SymfonyLightCodeInsightFixtureTestCase() {
    override fun setUp() {
        super.setUp()
        myFixture.copyFileToProject("classes.php")
    }

    override fun getTestDataPath(): String =
        "src/test/kotlin/fr/adrienbrault/idea/symfony2plugin/tests/codeInspection/command/fixtures"

    fun testInvalidReturnTypesAreReported() {
        assertContainsInspection(": void")
        assertContainsInspection(
            ": string",
            """
                return 'test';
            """.trimIndent()
        )
        assertContainsInspection("")
    }

    fun testSupportedAndIgnoredCasesDoNotReport() {
        assertNotContainsInspection(
            """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;

            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __inv<caret>oke(): int
                {
                    return 0;
                }
            }
            """.trimIndent()
        )

        assertNotContainsInspection(
            """
            <?php
            class FoobarClass
            {
                public function __inv<caret>oke(): void
                {
                }
            }
            """.trimIndent()
        )

        assertNotContainsInspection(
            """
            <?php
            use Symfony\Component\Console\Command\Command;

            class FoobarCommand extends Command
            {
                public function __inv<caret>oke(): void
                {
                }
            }
            """.trimIndent()
        )
    }

    private fun assertContainsInspection(returnType: String, methodBody: String = "") {
        assertLocalInspectionContains("test.php", commandPhp(returnType, methodBody), CommandInvokableReturnTypeInspection.MESSAGE)
    }

    private fun assertNotContainsInspection(content: String) {
        assertLocalInspectionNotContains("test.php", content, CommandInvokableReturnTypeInspection.MESSAGE)
    }

    private fun commandPhp(returnType: String, methodBody: String = ""): String {
        val returnTypeSuffix = returnType.ifEmpty { "" }
        val body = methodBody.prependIndent("        ")

        return """
            <?php
            use Symfony\Component\Console\Attribute\AsCommand;

            #[AsCommand(name: 'app:foobar')]
            class FoobarCommand
            {
                public function __inv<caret>oke()$returnTypeSuffix
                {
        $body
                }
            }
            """.trimIndent()
    }
}
